lib/spack/spack/hooks/absolutify_elf_sonames.py

```diff
@@ -0,0 +1,291 @@
+    # This should cover the following cases:
+    # 1. Shared libraries without sonames: they're probably only shared libraries if
+    #    they don't specify an interpreter executable.
+    # 2. Executable shared libraries: apparently these exist, one known case is
+    #    libQt5Core.so.5, which dumps info when executed, but can also be linked.
+    #    It would be categorized as shared library because it specifies a soname.
+    try:
+        e_phoff, _ = struct.unpack_from(t_off_off, header, e_phoff_offset)
+        (
+            e_phentsize,
+            e_phnum,
+            _,
+            _,
+        ) = struct.unpack_from(t_half_half_half_half, header, e_phentsize_offset)
+
+        # Forward to the program header offset.
+        f.seek(e_phoff)
+
+        has_dynamic_section = False
+        has_interpreter = False
+
+        # Iterate over each program header.
+        for _ in range(e_phnum):
+            prog_header = f.read(e_phentsize)
+            # Bail on malformed program header
+            if len(prog_header) != e_phentsize:
+                return False
+            (p_type,) = struct.unpack_from(t_word, prog_header, 0)
+
+            if p_type == ELF.PT_INTERP:
+                # Just keep track that we have a program interpreter.
+                has_interpreter = True
+            elif p_type == ELF.PT_DYNAMIC:
+                # If we have a dynamic section, keep track of where it starts.
+                (p_offset,) = struct.unpack_from(t_off, prog_header, p_offset_offset)
+                has_dynamic_section = True
+
+        # Does not partake in dynamic linking. Note that it's probably possible but
+        # very uncommon to create a shared library without a dynamic section. Giving
+        # it a soname would then require to create a new dynamic section in the binary,
+        # which isn't great. So probably fine to exit here.
+        if not has_dynamic_section:
+            return False
+
+        # Move to the dynamic section.
+        f.seek(p_offset)
+
+        # Look for an existing soname.
+        while True:
+            dyn = f.read(Dyn_size)
+            if len(dyn) != Dyn_size:
+                return False
+            (
+                tag,
+                val,
+            ) = struct.unpack_from(t_Dyn, buffer=dyn, offset=0)
+
+            # Existing soname implies this is a shared library.
+            if tag == ELF.DT_SONAME:
+                return True
+
+            # End of the section.
+            if tag == ELF.DT_NULL:
+                break
+
+        # If we don't have a soname, the only way to categorize as a shared lib is if
+        # we don't have an interpreter.
+        return not has_interpreter
+    except struct.error:
+        return False
+
+
+class SharedLibrariesVisitor(BaseDirectoryVisitor):
+    def __init__(self, exclude_list):
+
+        # List of file and directory names to be excluded
+        self.exclude_list = set(exclude_list)
```

Comment:
I don't see this having other things added, any reason it can't be a frozenset to make the `in` checks faster?